File: P2_Complejidad/fibonacci.c
#include <stdio.h>
#include <time.h>

long long fibonacci(int n){
    if(n == 0) return 0;
    if(n == 1) return 1;
    return fibonacci(n - 1) + fibonacci(n - 2);
}

int main(){
    int n;
    clock_t ini, fin;
    double tiempo;
    int memoria;

    printf("========================================\n");
    printf(" ALGORITMO RECURSIVO DE FIBONACCI\n");
    printf("========================================\n");
    printf("Este algoritmo calcula el n-esimo numero\n");
    printf("de la sucesion de Fibonacci usando\n");
    printf("recursividad.\n\n");
    printf("Regla de Fibonacci:\n");
    printf("F(0) = 0\n");
    printf("F(1) = 1\n");
    printf("F(n) = F(n-1) + F(n-2)\n\n");
    printf("Nota: Este metodo es ineficiente para\n");
    printf("valores grandes de n debido a la\n");
    printf("repeticion de calculos.\n");
    printf("========================================\n\n");

    printf("Ingrese el valor de n: ");
    scanf("%d", &n);

    memoria = n * sizeof(long long);

    ini = clock();
    long long resultado = fibonacci(n);
    fin = clock();

    tiempo = (double)(fin - ini) / CLOCKS_PER_SEC;

    printf("\nRESULTADO:\n");
    printf("Fibonacci(%d) = %lld\n", n, resultado);

    printf("\nESTADISTICAS:\n");
    printf("Tiempo de ejecucion: %f segundos\n", tiempo);
    printf("Memoria aproximada usada: %d bytes\n", memoria);

    printf("\nTabla comparativa:\n");
    printf("n\tTiempo(s)\tMemoria(bytes)\n");
    printf("%d\t%f\t%d\n", n, tiempo, memoria);

    return 0;
}

File: P2_Complejidad/fibonacci.py
import time
import sys

def fibonacci(n):
    if n == 0:
        return 0
    if n == 1:
        return 1
    return fibonacci(n - 1) + fibonacci(n - 2)

print("=" * 40)
print(" ALGORITMO RECURSIVO DE FIBONACCI")
print("=" * 40)
print("Este algoritmo calcula el n-esimo numero")
print("de la sucesion de Fibonacci usando")
print("recursividad.\n")
print("Regla de Fibonacci:")
print("F(0) = 0")
print("F(1) = 1")
print("F(n) = F(n-1) + F(n-2)\n")
print("Nota: Este metodo es ineficiente para")
print("valores grandes de n debido a la")
print("repeticion de calculos.")
print("=" * 40)

n = int(input("\nIngrese el valor de n: "))

memoria = n * sys.getsizeof(0)

inicio = time.perf_counter()
resultado = fibonacci(n)
fin = time.perf_counter()

tiempo = fin - inicio

print("\nRESULTADO:")
print(f"Fibonacci({n}) = {resultado}")

print("\nESTADISTICAS:")
print(f"Tiempo de ejecucion: {tiempo:.6f} segundos")
print(f"Memoria aproximada usada: {memoria} bytes")

print("\nTabla comparativa:")
print("n\tTiempo(s)\tMemoria(bytes)")
print(f"{n}\t{tiempo:.6f}\t{memoria}")